### DISCUSSION TITLE
Login not using the username provided by the POST request

### DISCUSSION THREAD

**Discussion by wolfsblu** [janko/rodauth-rails#191] [Q&A] [answered]
Hello! I'm having some strange issue when trying to call the login endpoint via POST with a JSON body. To be precise, when I make this call via curl or similar:
```
POST http://example.com/auth/login HTTP/1.1
Content-Type: application/json
{
    "username": "test",
    "password": "test"
}
```
I get the following response:
```
401 Unauthorized 
{
  "field-error": [
    "username",
    "no matching login"
  ],
  "error": "There was an error logging in"
}
```
And a look at the rails output reveals these SQL queries:
```
Started POST "/auth/login" for example.com at 2023-01-22 17:04:06 +0100
Processing by RodauthApp#call as HTML
  Parameters: {"username"=>"test", "password"=>"[FILTERED]"}
  Sequel (0.4ms)  SELECT version()
  ↳ app/misc/rodauth_app.rb:15:in `block (2 levels) in <class:RodauthApp>'
  Sequel (0.3ms)  SELECT version()
  ↳ app/misc/rodauth_app.rb:15:in `block (2 levels) in <class:RodauthApp>'
  Sequel (0.8ms)  SELECT * FROM `auth_users` WHERE ((`name` = '') AND (`status` IN (1, 2))) LIMIT 1
  ↳ app/misc/rodauth_app.rb:15:in `block (2 levels) in <class:RodauthApp>'
Completed 401 Unauthorized in 107ms (ActiveRecord: 3.7ms | Allocations: 13464)
```
Especially this line is interesting:
```
Sequel (0.8ms)  SELECT * FROM `auth_users` WHERE ((`name` = '') AND (`status` IN (1, 2))) LIMIT 1
```
It seems the credentials passed to the controller are not handed to rodauth somehow. I'm not sure whether I misconfigured my app or if there is some other issue at play here, but my configuration looks as follows:
```
class RodauthMain < Rodauth::Rails::Auth
  configure do
    enable :create_account, :verify_account, :verify_account_grace_period,
      :login, :logout, :remember, :json,
      :reset_password, :change_password, :change_password_notify,
      :change_login, :verify_login_change, :close_account

    accounts_table :auth_users 
    rails_account_model { Auth::User }

    login_column :name
    login_uses_email? false
    login_param :username

    email_to do 
      account[:email]
    end
  end
end
```
And my user model is also straight forward:
```
class Auth::User < ApplicationRecord
    include Rodauth::Rails.model
    enum :status, unverified: 1, verified: 2, closed: 3
end
```
Does anyone have an idea what could be wrong here? I'm using Ruby 3.2.0, rodauth 2.26.1 and rodauth-rails 1.7.0

**Comment by janko** [comment, marked as answer]
I believe the issue might be in `login_param` being passed a symbol, where it probably needs to be a string. Roda uses Rack's default params, which don't provide indifferent access that Active Support does, so params need to be access via string keys. Could you try whether it works with `login_param "username"`?

**Reply by wolfsblu** [reply]
I tried to use strings instead of symbols;
```
class RodauthMain < Rodauth::Rails::Auth
  configure do
    ...
    login_column "name"
    login_uses_email? false
    login_param "username"
  end
end
```
but unfortunately with the same result:
```
Started POST "/auth/login" for example.com at 2023-01-22 19:28:12 +0100
Processing by RodauthApp#call as HTML
  Parameters: {"username"=>"test", "password"=>"[FILTERED]"}
  Sequel (0.2ms)  SELECT version()
  ↳ app/misc/rodauth_app.rb:14:in `block (2 levels) in <class:RodauthApp>'
  Sequel (0.2ms)  SELECT version()
  ↳ app/misc/rodauth_app.rb:14:in `block (2 levels) in <class:RodauthApp>'
  Sequel (0.4ms)  SELECT * FROM `auth_users` WHERE (('name' = '') AND (`status` IN (1, 2))) LIMIT 1
  ↳ app/misc/rodauth_app.rb:14:in `block (2 levels) in <class:RodauthApp>'
Completed 401 Unauthorized in 14ms (ActiveRecord: 0.7ms | Allocations: 3282)
```

**Reply by janko** [reply]
The `login_column` needs to stay a symbol, because Sequel requires column names to be symbols, that's unrelated to Rack request params. I was able to reproduce the issue in a fresh Rails app, and switching `login_param` to use a string fixed it:

```rb
    login_column :name
    login_uses_email? false
    login_param "username"
```
```sh
# before
❯ curl -H "Content-Type: application/json" localhost:3000/login -d '{"username":"test","password":"test"}'
{"field-error":["username","no matching login"],"error":"There was an error logging in"}

# after
❯ curl -H "Content-Type: application/json" localhost:3000/login -d '{"username":"test","password":"test"}'
{"success":"You have been logged in"}
```

I'm not sure why in your last output the param value is still `''`, I would expect it to be `'test'`. However, if you're still experiencing the issue, I would need a sample Rails app that reproduces it.

**Reply by wolfsblu** [reply]
@janko First off, thanks for your help! I've now managed to create a minimal setup for reproducing the issue:
https://github.com/wolfsblu/rodauth-login-issue

The steps are quite simple:

1. Clone the repo
2. Install the dependencies with `bundle install`
3. Run migrations with `rails db:migrate`
4. Start the dev server with `rails s`
5. (optional) Insert a sample user into the DB
6. Try to login via REST (see `requests/auth/login.rest`)
7. Notice the log output:
  ```
Started POST "/auth/login" for 192.168.56.1 at 2023-01-27 19:39:43 +0100
  Sequel (0.3ms)  SELECT NULL AS `nil` FROM `accounts` LIMIT 1
  ↳ app/misc/rodauth_app.rb:3:in `<class:RodauthApp>'
Processing by RodauthApp#call as HTML
  Parameters: {"email"=>"name@example.com", "password"=>"[FILTERED]"}
  Sequel (0.4ms)  SELECT * FROM `accounts` WHERE ((`email` = '') AND (`status` IN (1, 2))) LIMIT 1
  ↳ app/misc/rodauth_app.rb:12:in `block (2 levels) in <class:RodauthApp>'
Completed 401 Unauthorized in 3ms (ActiveRecord: 0.7ms | Allocations: 1139)
  ```

**Reply by janko** [reply]
In the output I see `email` being sent in request parameters, which won't work, it needs to be `login`, like you have in the REST example.

**Reply by wolfsblu** [reply]
My bad, I tried it now with the correct parameter names, but in the end I get the same result:
```
curl -X POST http://127.0.0.1:3001/auth/login \
> -H 'Content-Type: application/json' \
> -d '{"login": "user", "password": "secret"}'
```
And got this response
```
{"field-error":["login","no matching login"],"error":"There was an error logging in"}
```
With this log output
```
Started POST "/auth/login" for 127.0.0.1 at 2023-01-28 09:44:15 +0100
Processing by RodauthApp#call as */*
  Parameters: {"login"=>"user", "password"=>"[FILTERED]"}
  Sequel (0.3ms)  SELECT * FROM `accounts` WHERE ((`email` = '') AND (`status` IN (1, 2))) LIMIT 1
  ↳ app/misc/rodauth_app.rb:12:in `block (2 levels) in <class:RodauthApp>'
Completed 401 Unauthorized in 40ms (ActiveRecord: 0.3ms | Allocations: 3338)
```

**Reply by janko** [reply]
Thanks, I was able to reproduce the issue. It seems like the problem is in Falcon; the request parameters get parsed by Rails, but not by Roda, they end up being blank (both `login` & `password` are missing). My suspicion is that something is not rewinding `env["rack.input"]`, aka request body. When I switch to Puma, everything works.

I will try to come up with a minimal example that reproduces the bug, and open an issue in the Falcon repository.

**Reply by wolfsblu** [reply]
That's a relief, I thought I'm missing something obvious. Thank you very much for looking into it!

**Reply by janko** [reply]
I pushed a [fix](https://github.com/janko/rodauth-rails/commit/5e9e1164e67a65385ce2510fa53dc49c2e2ee7df) to the `main` branch, which should make rodauth-rails compatible with Falcon 🙂

**Reply by wolfsblu** [reply]
Wow, that was really fast, thanks again!
